File: CSESProjects.py
intervals = []
memo = {}
p = []


def dp(j):
    if j == 0:
        return 0
    elif j not in memo:
        memo[j] = max(intervals[j][2] + dp(p[j]), dp(j - 1))
    return memo[j]


def main():
    global p
    n = int(input())
    for i in range(n):
        a, b, w = input().split()
        intervals.append((int(a), int(b), int(w)))
    intervals.sort(key=lambda x: x[1])
    p = [0] * n
    i = 0
    j = 0
    while j < n:
        while intervals[i + 1][1] <= intervals[j][0]:
            i += 1
        p[j] = i
        j += 1
    print(dp(n - 1))


main()

File: cses-projects.cpp
// https://cses.fi/problemset/task/1140
#include <iostream>
#include <vector>
#include <algorithm>

#define start first
#define finish second.first
#define weight second.second

using namespace std;

bool compare_finish(pair<int, pair<int, int>> a, pair<int, pair<int, int>> b)
{
    if (a.first != b.first)
        return a.first < b.first;
    return a.start < b.start;
}

int main()
{
    int n;
    cin >> n;

    vector<pair<int, pair<int, int>>> intervals(n);

    int s, f, w;
    for (int i = 0; i < n; ++i)
    {
        cin >> s >> f >> w;
        intervals[i] = {s, {f, w}};
    }
    sort(intervals.begin(), intervals.end(), compare_finish);

    vector<int> p(n);
    for (int i = -1, j = 0; j < n; ++j)
    {
        while (intervals[i + 1].finish < intervals[j].start)
            ++i;

        p[j] = i;
    }

    vector<long long> dp(n + 1);
    dp[0] = 0;
    for (int i = 0; i < n; ++i)
    {
        dp[i + 1] = max(dp[i], intervals[i].weight + dp[p[i] + 1]);
    }
    cout << dp[n] << endl;
}